doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+    $ INSTALL_LOCATION="/opt"
+    $ GITHUB_USERNAME="new_mne_contributor"
 
-Importing
-^^^^^^^^^
-* Import modules in this order:
-  1. builtin
-  2. standard scientific (``numpy as np``, ``scipy`` submodules)
-  3. others
-  4. mne imports (relative within the MNE module, absolute in the examples)
-* Imports for ``matplotlib`` and optional modules (``sklearn``, and ``pandas``, etc.) within the MNE module should be nested (i.e., within a function or method, not at the top of a file).
+Then make a local clone of your remote fork ("origin"):
 
-Vizualization
-^^^^^^^^^^^^^
-* Add public functions to the :mod:`mne.viz` package and use these in the corresponding methods.
-* All visualization functions must accept a ``show`` parameter and return a ``fig`` handle.
-* Use ``RdBu_r`` colormap for signed data with a meaningful middle (zero-point) and ``Reds`` otherwise in visualization functions and examples.
+.. code-block:: console
 
-Return types
-^^^^^^^^^^^^
-* Methods should modify inplace and return ``self``, functions should return copies (where applicable).
+    $ cd $INSTALL_LOCATION
+    $ git clone https://github.com/$GITHUB_USERNAME/mne-python.git
+
+Finally, set up a link between your local clone and the official repository
+("upstream"):
+
+.. code-block:: console
+
+    $ cd mne-python
+    $ git remote add upstream git://github.com/mne-tools/mne-python.git
+    $ git fetch --all
+
+.. note::
+
+    We use ``git://`` instead of ``https://`` in the address for the official
+    "upstream" remote repository. ``git://`` addresses are read-only, which
+    means you can *pull* the official repository into your local copy (in order
+    to stay up-to-date with changes made by other contributors) but you cannot
+    *push* anything from your computer directly into the upstream remote.
+    Instead, you must push your changes to your own remote fork first, and then
+    create a pull request from your remote into the upstream remote (and even
+    then, your changes are not automatically accepted into the upstream; the
+    changes must be *approved* by maintainers and then *merged* into upstream).
+
+Next, use the `environment file`_ provided in the root of the MNE-Python
+repository to set up your local development environment. This will install all
+of the dependencies needed for running MNE-Python. The following commands
+create the conda environment:
+
+.. code-block:: console
+
+    $ PREFERRED_ENVIRONMENT_NAME="mnedev"
+    $ conda env create -n $PREFERRED_ENVIRONMENT_NAME -f environment.yml
+    $ conda activate $PREFERRED_ENVIRONMENT_NAME
+
+.. note::
+
+    When using the environment file to install with Anaconda or Miniconda, the
+    name of the environment (``mne``) is built into the environment file
+    itself, but can be changed on the command line with the ``-n`` flag (as
+    shown above). This is helpful when maintaining separate environments for
+    stable and development versions of MNE-Python, or when using the
```

Comment:
Ahh this is the justification I missed above. I think it should come first, before telling users to do this other-env step. Say why it's done, then do it (not the reverse).